camus-api/src/main/java/com/linkedin/camus/schemaregistry/CachedSchemaRegistry.java

```diff
@@ -1,87 +1,156 @@
-			result = prime * result + ((id == null) ? 0 : id.hashCode());
-			result = prime * result + ((topic == null) ? 0 : topic.hashCode());
-			return result;
-		}
-
-		@Override
-		public boolean equals(Object obj) {
-			if (this == obj)
-				return true;
-			if (obj == null)
-				return false;
-			if (getClass() != obj.getClass())
-				return false;
-			CachedSchemaTuple other = (CachedSchemaTuple) obj;
-			if (id == null) {
-				if (other.id != null)
-					return false;
-			} else if (!id.equals(other.id))
-				return false;
-			if (topic == null) {
-				if (other.topic != null)
-					return false;
-			} else if (!topic.equals(other.topic))
-				return false;
-			return true;
-		}
-
-		@Override
-		public String toString() {
-			return "CachedSchemaTuple [topic=" + topic + ", id=" + id + "]";
-		}
-	}
+  private static final String GET_SCHEMA_BY_ID_MAX_RETIRES = "get.schema.by.id.max.retries";
+  private static final String GET_SCHEMA_BY_ID_MAX_RETIRES_DEFAULT = "3";
+  private static final String GET_SCHEMA_BY_ID_MIN_INTERVAL_SECONDS = "get.schema.by.id.min.interval.seconds";
+  private static final String GET_SCHEMA_BY_ID_MIN_INTERVAL_SECONDS_DEFAULT = "1";
+
+  private final SchemaRegistry<S> registry;
+  private final ConcurrentHashMap<CachedSchemaTuple, S> cachedById;
+  private final ConcurrentHashMap<String, SchemaDetails<S>> cachedLatest;
+  private int getByIdMaxRetries;
+  private int getByIdMinIntervalSeconds;
+  private Map<String, FailedFetchHistory> failedFetchHistories;
+
+  private class FailedFetchHistory {
+    private int numOfAttempts;
+    private long previousAttemptTime;
+
+    private FailedFetchHistory(int numOfnumOfAttempts, long previousAttemptTime) {
+      this.numOfAttempts = numOfnumOfAttempts;
+      this.previousAttemptTime = previousAttemptTime;
+    }
+  }
+
+  public void init(Properties props) {
+    this.getByIdMaxRetries =
+        Integer.parseInt(props.getProperty(GET_SCHEMA_BY_ID_MAX_RETIRES, GET_SCHEMA_BY_ID_MAX_RETIRES_DEFAULT));
+    this.getByIdMinIntervalSeconds =
+        Integer.parseInt(props.getProperty(GET_SCHEMA_BY_ID_MIN_INTERVAL_SECONDS,
+            GET_SCHEMA_BY_ID_MIN_INTERVAL_SECONDS_DEFAULT));
+  }
+
+  public CachedSchemaRegistry(SchemaRegistry<S> registry, Properties props) {
+    this.registry = registry;
+    this.cachedById = new ConcurrentHashMap<CachedSchemaTuple, S>();
+    this.cachedLatest = new ConcurrentHashMap<String, SchemaDetails<S>>();
+    this.failedFetchHistories = new HashMap<String, FailedFetchHistory>();
+    this.init(props);
+  }
+
+  public String register(String topic, S schema) {
+    return registry.register(topic, schema);
+  }
+
+  public S getSchemaByID(String topic, String id) {
+    CachedSchemaTuple cacheKey = new CachedSchemaTuple(topic, id);
+    S schema = cachedById.get(cacheKey);
+    if (schema == null) {
+      synchronized (this) {
+        if (shouldFetchFromSchemaRegistry(id)) {
+          schema = fetchFromSchemaRegistry(topic, id);
```

Comment:
Because if a thread starts to execute line 61, other thread should get false from line 60. So I think we need to put 60 and 61 in one synchronized block.